File: backend/src/magi/api/routers/schedules.py
"""Scheduler task management REST endpoints."""

from __future__ import annotations

import time
from dataclasses import asdict
from pathlib import Path
from typing import Any, Optional

from fastapi import APIRouter, Body, HTTPException, Query, Response, status
from pydantic import BaseModel, Field

from ... import i18n as core_i18n
from ...core.runtime_bindings import require_scheduler_service
from ...scheduler.contracts import (
    ScheduleDefinition,
    ScheduledTargetType,
    TriggerDefinition,
    TriggerType,
)
from ...scheduler.repository import ScheduleRepository
from ...utils.runtime import get_runtime_paths

schedules_router = APIRouter()


class ScheduleTriggerBody(BaseModel):
    trigger_type: TriggerType
    config: dict[str, Any] = Field(default_factory=dict)


class ScheduleUpdateBody(BaseModel):
    trigger: Optional[ScheduleTriggerBody] = None
    target_payload: Optional[dict[str, Any]] = None
    enabled: Optional[bool] = None
    metadata: Optional[dict[str, Any]] = None


class ScheduleCreateBody(BaseModel):
    schedule_id: str = Field(min_length=1, max_length=300)
    target_type: ScheduledTargetType
    target_key: str = Field(min_length=1, max_length=300)
    trigger: ScheduleTriggerBody
    target_payload: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True
    metadata: dict[str, Any] = Field(default_factory=dict)


class ScheduleRunBody(BaseModel):
    """Optional body for ``POST /schedules/{id}/run`` — generic manual-trigger params.

    ``override_params`` is shallow-merged into the schedule's stored
    ``target_payload`` for this execution only; the DB row is not mutated.
    Handlers opt in by reading ``context.schedule.target_payload``.

    Example: ``{"override_params": {"days": 7}}`` to ask the diary handler
    to backfill 7 days in one trigger.
    """

    override_params: dict[str, Any] = Field(default_factory=dict)


class ActivityCancelBody(BaseModel):
    reason: str = Field(default="cancelled_by_user", max_length=200)


def _repository() -> ScheduleRepository:
    runtime_paths = get_runtime_paths()
    scheduler_db_path = getattr(
        runtime_paths,
        "scheduler_db_path",
        Path(runtime_paths.base_dir) / "runtime" / "scheduler.db",
    )
    return ScheduleRepository(scheduler_db_path)


def _serialize_state(state) -> dict[str, Any]:
    return {
        "target_type": state.target_type.value,
        "target_key": state.target_key,
        "running": state.running,
        "last_run_at": state.last_run_at,
        "last_success_at": state.last_success_at,
        "last_error": state.last_error,
        "last_cursor": state.last_cursor,
        "watermark_ts": state.watermark_ts,
        "next_run_at": state.next_run_at,
        "scheduler_job_id": state.scheduler_job_id,
        "updated_at": state.updated_at,
        "stats": state.stats,
    }


def _serialize_schedule(schedule: ScheduleDefinition, state=None) -> dict[str, Any]:
    metadata = dict(schedule.metadata or {})
    editable = schedule.target_type is not ScheduledTargetType.SOURCE_SYNC
    if schedule.target_type is ScheduledTargetType.SOURCE_SYNC:
        owner_kind = "source_settings"
    elif schedule.target_type is ScheduledTargetType.USER_AGENT_TASK:
        owner_kind = "agent_created"
    else:
        owner_kind = "system"
    payload = dict(schedule.target_payload or {})
    source_name = payload.get("source_type") or metadata.get("source_type")
    settings_link = (
        {"section": "timeline", "source_name": source_name}
        if schedule.target_type is ScheduledTargetType.SOURCE_SYNC and source_name
        else None
    )
    return {
        "schedule_id": schedule.schedule_id,
        "target_type": schedule.target_type.value,
        "target_key": schedule.target_key,
        "trigger": {
            "trigger_type": schedule.trigger.trigger_type.value,
            "config": schedule.trigger.config,
        },
        "target_payload": payload,
        "enabled": schedule.enabled,
        "metadata": metadata,
        "job_id": schedule.job_id,
        "editable": editable,
        "owner_kind": owner_kind,
        "settings_link": settings_link,
        "target_state": _serialize_state(state) if state is not None else None,
    }


def _schedule_title(schedule: ScheduleDefinition) -> str:
    metadata = schedule.metadata or {}
    payload = schedule.target_payload or {}
    for key in ("display_name", "title", "source_type", "plugin_id"):
        value = metadata.get(key) or payload.get(key)
        if value:
            return str(value)
    return schedule.schedule_id


def _source_job_activity(job: dict[str, Any], schedule: ScheduleDefinition | None) -> dict[str, Any]:
    title = _schedule_title(schedule) if schedule is not None else str(job["source_type"])
    queued = str(job["status"]) == "queued"
    return {
        "activity_id": f"source_job:{job['job_id']}",
        "schedule_id": job["schedule_id"],
        "title": title,
        "target_type": job["target_type"],
        "target_key": job["target_key"],
        "status": job["status"],
        "planned_at": job["created_at"],
        "started_at": job["started_at"],
        "duration_ms": None,
        "cancellable": queued,
        "cancel_kind": "source_sync_job" if queued else None,
        "error": job["error"],
    }


async def _source_job_activities(
    repository: ScheduleRepository,
    *,
    schedules: list[ScheduleDefinition],
    limit: int,
) -> list[dict[str, Any]]:
    schedule_by_id = {schedule.schedule_id: schedule for schedule in schedules}
    return [
        _source_job_activity(
            job,
            schedule_by_id.get(str(job["schedule_id"])),
        )
        for job in await repository.list_outstanding_source_sync_jobs(limit=limit)
    ]


def _running_schedule_activity(schedule: ScheduleDefinition, state: Any) -> dict[str, Any]:
    return {
        "activity_id": f"target:{schedule.target_type.value}:{schedule.target_key}",
        "schedule_id": schedule.schedule_id,
        "title": _schedule_title(schedule),
        "target_type": schedule.target_type.value,
        "target_key": schedule.target_key,
        "status": "running",
        "planned_at": None,
        "started_at": state.last_run_at,
        "duration_ms": (
            max(0.0, (time.time() - state.last_run_at) * 1000.0)
            if state.last_run_at
            else None
        ),
        "cancellable": False,
        "cancel_kind": None,
        "error": state.last_error,
    }


def _upcoming_schedule_activity(schedule: ScheduleDefinition, state: Any) -> dict[str, Any]:
    return {
        "activity_id": f"upcoming:{schedule.schedule_id}",
        "schedule_id": schedule.schedule_id,
        "title": _schedule_title(schedule),
        "target_type": schedule.target_type.value,
        "target_key": schedule.target_key,
        "status": "upcoming",
        "planned_at": state.next_run_at,
        "started_at": None,
        "duration_ms": None,
        "cancellable": False,
        "cancel_kind": None,
        "error": state.last_error,
    }


async def _schedule_state_activities(
    repository: ScheduleRepository,
    schedules: list[ScheduleDefinition],
) -> list[dict[str, Any]]:
    activities: list[dict[str, Any]] = []
    for schedule in schedules:
        state = await repository.get_schedule_runtime_state(schedule)
        if state.running and schedule.target_type is not ScheduledTargetType.SOURCE_SYNC:
            activities.append(_running_schedule_activity(schedule, state))
        if state.next_run_at is not None and not state.running:
            activities.append(_upcoming_schedule_activity(schedule, state))
    return activities


async def _get_schedule_or_404(
    repository: ScheduleRepository,
    schedule_id: str,
) -> ScheduleDefinition:
    schedule = await repository.get_schedule(schedule_id)
    if schedule is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=core_i18n.t("schedules.errors.not_found", fallback="Schedule not found"),
        )
    return schedule


def _scheduler_service_or_503():
    try:
        return require_scheduler_service()
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=core_i18n.t(
                "schedules.errors.service_unavailable",
                fallback="Scheduler service is not available",
            ),
        ) from exc


def _raise_schedule_run_error(message: str) -> None:
    if message == "schedule_not_found":
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=core_i18n.t("schedules.errors.not_found", fallback="Schedule not found"),
        )
    if message == "target_busy":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=core_i18n.t("schedules.errors.target_busy", fallback="Schedule target is busy"),
        )
    raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=message)


@schedules_router.get("")
async def list_schedules(
    enabled_only: bool = Query(default=False),
) -> dict[str, Any]:
    repository = _repository()
    await repository.initialize()
    schedules = await repository.list_schedules(enabled_only=enabled_only)
    items = []
    for schedule in schedules:
        state = await repository.get_schedule_runtime_state(schedule)
        items.append(_serialize_schedule(schedule, state))
    return {"schedules": items}


@schedules_router.post("", status_code=status.HTTP_201_CREATED)
async def create_schedule(body: ScheduleCreateBody) -> dict[str, Any]:
    schedule = ScheduleDefinition(
        schedule_id=body.schedule_id,
        target_type=body.target_type,
        target_key=body.target_key,
        trigger=TriggerDefinition(body.trigger.trigger_type, dict(body.trigger.config)),
        target_payload=dict(body.target_payload),
        enabled=body.enabled,
        metadata=dict(body.metadata),
        job_id=body.schedule_id,
    )
    repository = _repository()
    await repository.initialize()
    try:
        scheduler_service = require_scheduler_service()
    except RuntimeError:
        await repository.upsert_schedule(schedule)
        saved = await repository.get_schedule(schedule.schedule_id)
    else:
        saved = await scheduler_service.schedule(schedule)
    state = await repository.get_schedule_runtime_state(saved or schedule)
    return {"schedule": _serialize_schedule(saved or schedule, state)}


@schedules_router.get("/activity")
async def list_schedule_activity(
    limit: int = Query(default=100, ge=1, le=300),
) -> dict[str, Any]:
    repository = _repository()
    await repository.initialize()
    schedules = await repository.list_schedules(enabled_only=True)
    activities = await _source_job_activities(repository, schedules=schedules, limit=limit)
    activities.extend(await _schedule_state_activities(repository, schedules))
    activities.sort(key=lambda item: (item["status"] != "running", item.get("planned_at") or 0))
    return {"activities": activities[:limit]}


@schedules_router.patch("/{schedule_id}")
async def update_schedule(schedule_id: str, body: ScheduleUpdateBody) -> dict[str, Any]:
    repository = _repository()
    await repository.initialize()
    existing = await repository.get_schedule(schedule_id)
    if existing is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=core_i18n.t("schedules.errors.not_found", fallback="Schedule not found"),
        )
    if existing.target_type is ScheduledTargetType.SOURCE_SYNC:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=core_i18n.t(
                "schedules.errors.source_schedule_settings_only",
                fallback="Source schedules must be updated from source settings",
            ),
        )

    next_schedule = ScheduleDefinition(
        schedule_id=existing.schedule_id,
        target_type=existing.target_type,
        target_key=existing.target_key,
        trigger=(
            TriggerDefinition(body.trigger.trigger_type, dict(body.trigger.config))
            if body.trigger is not None
            else existing.trigger
        ),
        target_payload=(
            dict(body.target_payload)
            if body.target_payload is not None
            else dict(existing.target_payload)
        ),
        enabled=body.enabled if body.enabled is not None else existing.enabled,
        metadata=dict(body.metadata) if body.metadata is not None else dict(existing.metadata),
        job_id=existing.job_id,
    )
    try:
        scheduler_service = require_scheduler_service()
    except RuntimeError:
        await repository.upsert_schedule(next_schedule)
        saved = await repository.get_schedule(schedule_id)
    else:
        saved = await scheduler_service.schedule(next_schedule)
    state = await repository.get_schedule_runtime_state(saved or next_schedule)
    return {"schedule": _serialize_schedule(saved or next_schedule, state)}


@schedules_router.get("/{schedule_id}")
async def get_schedule(schedule_id: str) -> dict[str, Any]:
    repository = _repository()
    await repository.initialize()
    schedule = await repository.get_schedule(schedule_id)
    if schedule is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=core_i18n.t("schedules.errors.not_found", fallback="Schedule not found"),
        )
    state = await repository.get_schedule_runtime_state(schedule)
    return {"schedule": _serialize_schedule(schedule, state)}


@schedules_router.delete("/{schedule_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_schedule(schedule_id: str) -> Response:
    repository = _repository()
    await repository.initialize()
    existing = await repository.get_schedule(schedule_id)
    if existing is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=core_i18n.t("schedules.errors.not_found", fallback="Schedule not found"),
        )
    try:
        scheduler_service = require_scheduler_service()
    except RuntimeError:
        await repository.clear_target_schedule_binding(existing.target_type, existing.target_key)
        await repository.delete_schedule(schedule_id)
    else:
        await scheduler_service.unschedule(
            schedule_id,
            target_type=existing.target_type,
            target_key=existing.target_key,
        )
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@schedules_router.post("/{schedule_id}/run")
async def run_schedule_now(
    schedule_id: str,
    body: ScheduleRunBody | None = Body(default=None),
) -> dict[str, Any]:
    repository = _repository()
    await repository.initialize()
    existing = await _get_schedule_or_404(repository, schedule_id)
    scheduler_service = _scheduler_service_or_503()

    override_params = (body.override_params if body is not None else {}) or {}
    # Fire-and-forget: returns after setup (lock + execution record); the
    # handler runs in a background task. The HTTP request used to hang for
    # the full duration of the handler — for diary backfill that's minutes
    # of dead UI. Now the response is sub-second; status moves through the
    # /schedules/activity feed.
    result = await scheduler_service.execute_schedule_async(
        schedule_id, manual=True, override_payload=override_params or None,
    )
    if not result.success:
        _raise_schedule_run_error(result.message)

    saved = await repository.get_schedule(schedule_id)
    state = await repository.get_schedule_runtime_state(saved or existing)
    return {
        "schedule": _serialize_schedule(saved or existing, state),
        "result": asdict(result),
    }


@schedules_router.post("/activity/{activity_id}/cancel")
async def cancel_schedule_activity(
    activity_id: str,
    body: Optional[ActivityCancelBody] = None,
) -> dict[str, Any]:
    repository = _repository()
    await repository.initialize()
    reason = (body.reason if body is not None else "cancelled_by_user") or "cancelled_by_user"
    if not activity_id.startswith("source_job:"):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=core_i18n.t("schedules.errors.activity_not_cancellable", fallback="Activity is not cancellable"),
        )
    job_id = activity_id.split(":", 1)[1]
    job = await repository.cancel_queued_source_sync_job(job_id, reason=reason)
    if job is None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=core_i18n.t("schedules.errors.activity_not_cancellable", fallback="Activity is not cancellable"),
        )
    return {
        "activity": {
            "activity_id": activity_id,
            "status": job["status"],
            "job_id": job["job_id"],
        }
    }


__all__ = ["schedules_router"]

File: frontend/src/api/modules/schedules.ts
import { api, unwrapGatewayPayload } from '../client';
import { z } from 'zod';

export type ScheduleTargetType =
  | 'source_sync'
  | 'memory_l2_maintenance'
  | 'memory_l3_summary'
  | 'user_agent_task'
  | string;

export type ScheduleTriggerType = 'once' | 'interval' | 'cron';

export interface ScheduleTriggerDTO {
  trigger_type: ScheduleTriggerType;
  config: Record<string, unknown>;
}

export interface ScheduleTargetStateDTO {
  target_type: ScheduleTargetType;
  target_key: string;
  running: boolean;
  last_run_at?: number | null;
  last_success_at?: number | null;
  last_error?: string | null;
  last_cursor?: string | null;
  watermark_ts?: number | null;
  next_run_at?: number | null;
  scheduler_job_id?: string | null;
  updated_at?: number | null;
  stats?: Record<string, unknown>;
}

export interface ScheduleSettingsLinkDTO {
  section: 'timeline' | 'memory' | string;
  source_name?: string | null;
}

export interface ScheduleDTO {
  revision: number;
  schedule_id: string;
  target_type: ScheduleTargetType;
  target_key: string;
  trigger: ScheduleTriggerDTO;
  target_payload: Record<string, unknown>;
  enabled: boolean;
  metadata: Record<string, unknown>;
  job_id?: string | null;
  editable?: boolean;
  owner_kind?: 'source_settings' | 'system' | 'user' | string;
  settings_link?: ScheduleSettingsLinkDTO | null;
  target_state?: ScheduleTargetStateDTO | null;
}

export interface ListSchedulesResponse {
  schedules: ScheduleDTO[];
}

export type ScheduleActivityStatus =
  | 'running'
  | 'queued'
  | 'upcoming'
  | 'succeeded'
  | 'failed'
  | 'cancelled'
  | string;

export interface ScheduleActivityDTO {
  activity_id: string;
  schedule_id: string;
  title?: string | null;
  target_type: ScheduleTargetType;
  target_key: string;
  status: ScheduleActivityStatus;
  planned_at?: number | null;
  started_at?: number | null;
  finished_at?: number | null;
  duration_ms?: number | null;
  cancellable: boolean;
  cancel_kind?: 'source_sync_job' | string | null;
  error?: string | null;
  background_task_id?: string | null;
  result_message?: string | null;
  stats?: Record<string, unknown>;
  manual?: boolean;
}

export interface ListScheduleActivityResponse {
  activities: ScheduleActivityDTO[];
  total?: number;
  /**
   * Counts of all activity rows in the requested time window, grouped by
   * raw `target_type`. Independent of any category/status filter applied —
   * intended for driving filter-chip counters that need to remain stable
   * when the user clicks a chip.
   */
  target_type_counts?: Record<string, number>;
  /**
   * Counts of all activity rows in the requested time window, grouped by
   * display status (succeeded / failed / running / queued / cancelled).
   * Same semantics as `target_type_counts`.
   */
  status_counts?: Record<string, number>;
}

export interface ListActivityParams {
  sinceSeconds?: number;
  untilSeconds?: number;
  limit?: number;
  offset?: number;
  targetTypes?: string[];
  statuses?: string[];
}

export interface CreateScheduleRequest {
  schedule_id: string;
  display_name: string;
  prompt: string;
  trigger: ScheduleTriggerDTO;
  enabled: boolean;
}

export interface CreateScheduleResponse {
  schedule: ScheduleDTO;
}

export interface UpdateScheduleRequest {
  revision: number;
  trigger?: ScheduleTriggerDTO;
  target_payload?: Record<string, unknown>;
  enabled?: boolean;
  metadata?: Record<string, unknown>;
}

export interface UpdateScheduleResponse {
  schedule: ScheduleDTO;
}

export interface ScheduledExecutionResultDTO {
  success: boolean;
  message?: string;
  next_cursor?: string | null;
  watermark_ts?: number | null;
  stats?: Record<string, unknown>;
}

export interface RunScheduleResponse {
  schedule: ScheduleDTO;
  result: ScheduledExecutionResultDTO;
}

export interface CancelScheduleActivityResponse {
  activity: {
    activity_id: string;
    status: string;
    job_id?: string;
  };
}

const scheduleSchema = z.object({
  schedule_id: z.string(), revision: z.number().finite().positive(), target_type: z.string(),
  target_key: z.string(), trigger: z.object({ trigger_type: z.enum(['once', 'interval', 'cron']), config: z.record(z.string(), z.unknown()) }),
  target_payload: z.record(z.string(), z.unknown()), enabled: z.boolean(), metadata: z.record(z.string(), z.unknown()),
  job_id: z.string().nullish(), editable: z.boolean().optional(), owner_kind: z.string().optional(),
  settings_link: z.object({ section: z.string(), source_name: z.string().nullish() }).nullish(),
  target_state: z.object({
    target_type: z.string(), target_key: z.string(), running: z.boolean(),
    last_run_at: z.number().nullish(), last_success_at: z.number().nullish(), last_error: z.string().nullish(),
    last_cursor: z.string().nullish(), watermark_ts: z.number().nullish(), next_run_at: z.number().nullish(),
    scheduler_job_id: z.string().nullish(), updated_at: z.number().nullish(), stats: z.record(z.string(), z.unknown()).optional(),
  }).nullish(),
});
const parseSchedule = (value: unknown): ScheduleDTO => {
  // Narrow the hand-owned schedule boundary before the UI uses its revision.
  const parsed = scheduleSchema.parse(value);
  return parsed;
};

export const schedulesApi = {
  async get(scheduleId: string): Promise<ScheduleDTO> {
    const response = await api.get<{ schedule: unknown }>(`/schedules/${encodeURIComponent(scheduleId)}`);
    return parseSchedule(unwrapGatewayPayload(response).schedule);
  },
  async list(params: { enabledOnly?: boolean } = {}): Promise<ListSchedulesResponse> {
    const search = new URLSearchParams();
    if (params.enabledOnly !== undefined) {
      search.set('enabled_only', String(params.enabledOnly));
    }
    const query = search.toString();
    const response = await api.get<ListSchedulesResponse>(`/schedules${query ? `?${query}` : ''}`);
    const payload = unwrapGatewayPayload(response);
    return { schedules: payload.schedules.map(parseSchedule) };
  },

  async listActivity(params: ListActivityParams = {}): Promise<ListScheduleActivityResponse> {
    const search = new URLSearchParams();
    if (params.sinceSeconds !== undefined) search.set('since', String(params.sinceSeconds));
    if (params.untilSeconds !== undefined) search.set('until', String(params.untilSeconds));
    if (params.limit !== undefined) search.set('limit', String(params.limit));
    if (params.offset !== undefined) search.set('offset', String(params.offset));
    (params.targetTypes ?? []).forEach((t) => search.append('target_types', t));
    (params.statuses ?? []).forEach((s) => search.append('statuses', s));
    const query = search.toString();
    const response = await api.get<ListScheduleActivityResponse>(
      `/schedules/activity${query ? `?${query}` : ''}`,
    );
    return unwrapGatewayPayload(response);
  },

  async create(body: CreateScheduleRequest): Promise<CreateScheduleResponse> {
    const wireBody = {
      schedule_id: body.schedule_id,
      target_type: 'user_agent_task' as const,
      target_key: body.schedule_id,
      trigger: body.trigger,
      target_payload: { prompt: body.prompt, kind: 'agent_task' },
      metadata: { display_name: body.display_name, target_kind: 'agent_task' },
      enabled: body.enabled,
    };
    const response = await api.post<CreateScheduleResponse>('/schedules', wireBody);
    return { schedule: parseSchedule(unwrapGatewayPayload(response).schedule) };
  },

  async update(scheduleId: string, body: UpdateScheduleRequest): Promise<UpdateScheduleResponse> {
    const response = await api.patch<UpdateScheduleResponse>(
      `/schedules/${encodeURIComponent(scheduleId)}`,
      body,
    );
    return { schedule: parseSchedule(unwrapGatewayPayload(response).schedule) };
  },

  async remove(scheduleId: string, revision: number): Promise<void> {
    await api.delete(`/schedules/${encodeURIComponent(scheduleId)}?revision=${encodeURIComponent(revision)}`);
  },

  /**
   * Manually trigger a schedule. ``overrideParams``, when provided, is
   * shallow-merged on top of the schedule's stored target_payload for this
   * single execution (the DB row is not mutated). Handlers opt in by
   * reading ``context.schedule.target_payload``.
   *
   * Example: ``run("timeline_diary_narrative", { days: 7 })`` to backfill
   * 7 days of diary narratives in one trigger.
   */
  async run(
    scheduleId: string,
    overrideParams?: Record<string, unknown>,
  ): Promise<RunScheduleResponse> {
    const response = await api.post<RunScheduleResponse>(
      `/schedules/${encodeURIComponent(scheduleId)}/run`,
      overrideParams && Object.keys(overrideParams).length > 0
        ? { override_params: overrideParams }
        : undefined,
    );
    return unwrapGatewayPayload(response);
  },

  async cancelActivity(activityId: string, reason?: string): Promise<CancelScheduleActivityResponse> {
    const response = await api.post<CancelScheduleActivityResponse>(
      `/schedules/activity/${encodeURIComponent(activityId)}/cancel`,
      reason ? { reason } : {},
    );
    return unwrapGatewayPayload(response);
  },
};
